tests/unit_tests/long_term_block_weight.cpp

```diff
@@ -0,0 +1,384 @@
+      ASSERT_TRUE(bc->update_next_cumulative_weight_limit());
+    }
+
+    // check the new values are the same as the old ones
+    for (int i = -2; i < std::min(add, remove); ++i)
+    {
+      ASSERT_EQ(bc->get_db().get_block_long_term_weight(h0 + i), old_ltbw[i + 2]);
+    }
+  }
+}
+
+TEST(long_term_block_weight, pop_invariant_random)
+{
+  PREFIX(10);
+
+  for (uint64_t h = 1; h < TEST_LONG_TERM_BLOCK_WEIGHT_WINDOW - 10; ++h)
+  {
+    size_t w = h < TEST_LONG_TERM_BLOCK_WEIGHT_WINDOW ? CRYPTONOTE_BLOCK_GRANTED_FULL_REWARD_ZONE_V5 : bc->get_current_cumulative_block_weight_limit();
+    uint64_t ltw = bc->get_next_long_term_block_weight(w);
+    bc->get_db().add_block(cryptonote::block(), w, ltw, h, h, {});
+    ASSERT_TRUE(bc->update_next_cumulative_weight_limit());
+  }
+
+  for (int n = 0; n < 1000; ++n)
+  {
+    // pop some blocks, then add some more
+    int remove = 1 + (n * 17) % 8;
+    int add = (n * 23) % 12;
+
+    // save long term block weights we're about to remove
+    uint64_t old_ltbw[16], h0 = bc->get_db().height() - remove - 1;
+    for (int i = -2; i < remove; ++i)
+    {
+      old_ltbw[i + 2] = bc->get_db().get_block_long_term_weight(h0 + i);
+    }
+
+    for (int i = 0; i < remove; ++i)
+    {
+      cryptonote::block b;
+      std::vector<cryptonote::transaction> txs;
+      bc->get_db().pop_block(b, txs);
+      ASSERT_TRUE(bc->update_next_cumulative_weight_limit());
+      const uint64_t effective_median = bc->get_current_cumulative_block_weight_median();
+      const uint64_t effective_limit = bc->get_current_cumulative_block_weight_limit();
+      ASSERT_TRUE(bc->update_next_cumulative_weight_limit());
+      ASSERT_EQ(effective_median, bc->get_current_cumulative_block_weight_median());
+      ASSERT_EQ(effective_limit, bc->get_current_cumulative_block_weight_limit());
+    }
+    for (int i = 0; i < add; ++i)
+    {
+      lcg_seed = bc->get_db().height();
+      uint32_t r = lcg();
+      size_t w = bc->get_db().height() < TEST_LONG_TERM_BLOCK_WEIGHT_WINDOW ? CRYPTONOTE_BLOCK_GRANTED_FULL_REWARD_ZONE_V5 : (r % bc->get_current_cumulative_block_weight_limit());
+      uint64_t ltw = bc->get_next_long_term_block_weight(w);
+      bc->get_db().add_block(cryptonote::block(), w, ltw, bc->get_db().height(), bc->get_db().height(), {});
+      ASSERT_TRUE(bc->update_next_cumulative_weight_limit());
+      const uint64_t effective_median = bc->get_current_cumulative_block_weight_median();
+      const uint64_t effective_limit = bc->get_current_cumulative_block_weight_limit();
+      ASSERT_TRUE(bc->update_next_cumulative_weight_limit());
+      ASSERT_EQ(effective_median, bc->get_current_cumulative_block_weight_median());
+      ASSERT_EQ(effective_limit, bc->get_current_cumulative_block_weight_limit());
+    }
+
+    // check the new values are the same as the old ones
+    for (int i = -2; i < std::min(add, remove); ++i)
+    {
+      ASSERT_EQ(bc->get_db().get_block_long_term_weight(h0 + i), old_ltbw[i + 2]);
```

Comment:
I don't see this holding necessarily. I've confirmed the issue by making the following change
```diff
-    int add = (n * 23) % 12;
+    int add = (n * 23) % 123;
```
which results in the following error:
```
[ RUN      ] long_term_block_weight.pop_invariant_random
/monero/tests/unit_tests/long_term_block_weight.cpp:329: Failure
Value of: old_ltbw[i + 2]
  Actual: 440230
Expected: bc->get_db().get_block_long_term_weight(h0 + i)
Which is: 440020
[  FAILED  ] long_term_block_weight.pop_invariant_random (4232 ms)
```

Reply:
No, it should always work :/

Reply:
Why do this test? The blocks added will have different (randomized) weights than the removed blocks, AIUI.